src/components/ApiQueryEditor/ApiQueryEditor.tsx

```diff
@@ -0,0 +1,118 @@
+    defaultValue?: string;
+}
+
+export interface ApiDS extends DataSourceWithBackend {
+    apiList: () => API[];
+    lookup?: (type: string) => Promise<SelectableValue[]>;
+}
+
+export type Props = {
+    query: ApiQuery;
+    onChange: (query: DataQuery) => void;
+    datasource: ApiDS;
+}
+
+export const ApiQueryEditor = (props: Props) => {
+    const { datasource, query } = props;
+    const defaultInputs: InputValue[] = [];
+    const apiList = datasource.apiList();
+    const api = apiList.find((a) => a.name === query.queryType) || {name: '', inputs: defaultInputs};
+    const [inputs, setInputs] = useState<InputValue[]>(api.inputs);
+
+    const onChangeSelect = (selected: SelectableValue<string>, i: number) => {
+        const val = {currentTarget: {value: selected.value!}} as unknown as FormEvent<HTMLInputElement>;
+        onChangeInput(val, i)
+    }
+
+    const onChangeInput = (v: FormEvent<HTMLInputElement>, i: number) => {
+        const updatedInputs = inputs?.map((input, idx) => {
+            if (i === idx) {
+                return {...input, value: v.currentTarget.value};
+            }
+            return {...input, value: value(query, idx)};
+        })
+        setInputs(updatedInputs);
+        const update = {...query, inputs: updatedInputs};
+        props.onChange(update);
+    }
+
+    const loadOptions = async (lookup: string) => {
+        return await datasource.lookup!(lookup);
+    }
+
+    return (
+        <div>
+            {api.inputs.map((iv: InputValue, i) => {
+                const val = value(query, i, iv.defaultValue);
+                return <div className="gf-form" key={i}>
+                        <InlineFormLabel tooltip={iv.description}>
+                            {pretty(iv.name)}
+                        </InlineFormLabel>
+                        {iv.choices &&  
+                            <Select
+                                allowCustomValue
+                                isClearable
+                                options={toSelectable(iv.choices)}
+                                width={40}
+                                value={val}
+                                onChange={(selected: SelectableValue<string>, _: ActionMeta) => onChangeSelect(selected, i)}
+                            />
+                        }
+                        {iv.lookup &&  
+                            <AsyncSelect
+                                loadOptions={async () => await loadOptions(iv.lookup!)}
+                                defaultOptions={true}
+                                allowCustomValue
+                                isClearable
+                                width={40}
+                                value={{label: val, value: val}}
+                                onChange={(selected: SelectableValue<string>, _: ActionMeta) => onChangeSelect(selected, i)}
+                            />
+                        }
+                        {!iv.choices && !iv.lookup &&  
```

Comment:
Seems like `lookup`, `choices` etc. are mutually exclusive and I'd suggest to rethink the `InputValue` shape. I'd suggest to have something like `type` which can have a limited number of options like `select` , `asyncselect`, `input`, etc. Maybe it can be some nested structure depending on the `type`.

The reason I am suggesting this is I think in this way it will be more future proof. Maybe in the future we would want to also render the options as a radio buttons, etc. And for such cases I think we should create API in a way that it's easily extendible. It might be hard to change the API when plugins start using it.